File: 384_shuffle_array/shuffle_array.cpp
/*
 * @Date: 2021-11-22 02:07:14
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-11-22 02:18:43
 */

#include <iostream>
#include <vector>

using namespace std;

class Solution {
 public:
  Solution(vector<int>& nums) {
    this->nums = nums;
    this->original.resize(nums.size());
    copy(nums.begin(), nums.end(), original.begin());
  }

  vector<int> reset() {
    copy(original.begin(), original.end(), nums.begin());
    return nums;
  }

  vector<int> shuffle() {
    int nums_size = nums.size();
    for (int i = 0; i < nums_size; ++i) {
      int j = i + rand() % (nums.size() - i);
      swap(nums[i], nums[j]);
    }
    return nums;
  }

 private:
  vector<int> nums;
  vector<int> original;
};

void print_vector(vector<int> nums) {
    int nums_size = nums.size();
  for (int i = 0; i < nums_size; ++i) {
    cout << nums[i] << " ";
  }
  cout << endl;
}

int main() {
  vector<int> nums = {1, 2, 3};
  Solution s(nums);
  print_vector(s.shuffle());
  print_vector(s.reset());
  print_vector(s.shuffle());
}

File: golang/384_shuffle_array/shuffle_array.go
/*
 * @Date: 2021-11-22 02:11:22
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-11-22 02:19:01
 */

package main

import (
	"fmt"
	"math/rand"
)

type Solution struct {
	nums, original []int
}

func Constructor(nums []int) Solution {
	return Solution{nums, append([]int(nil), nums...)}
}

func (s *Solution) Reset() []int {
	copy(s.nums, s.original)
	return s.nums
}

func (s *Solution) Shuffle() []int {
	n := len(s.nums)
	for i := range s.nums {
		j := i + rand.Intn(n-i)
		s.nums[i], s.nums[j] = s.nums[j], s.nums[i]
	}
	return s.nums
}

func main() {
	{
		s := Constructor([]int{1, 2, 3})
		fmt.Println(s.Shuffle())
		fmt.Println(s.Reset())
		fmt.Println(s.Shuffle())
	}
}
